mp4parse/src/lib.rs

```diff
@@ -1105,16 +1117,77 @@ fn read_flac_metadata<T: Read>(src: &mut BMFFBox<T>) -> Result<FLACMetadataBlock
     })
 }
 
-fn read_esds<T: Read>(src: &mut BMFFBox<T>) -> Result<AudioCodecSpecific> {
+fn read_esds<T: Read>(src: &mut BMFFBox<T>) -> Result<EsDescriptor> {
+    // Parsing DecoderConfig descriptor to get the codec type.
     let (_, _) = try!(read_fullbox_extra(src));
 
     let esds_size = src.head.size - src.head.offset - 4;
     if esds_size > BUF_SIZE_LIMIT {
         return Err(Error::InvalidData("esds box exceeds BUF_SIZE_LIMIT"));
     }
-    let esds = try!(read_buf(&mut src.content, esds_size as usize));
+    let esds_array = try!(read_buf(src, esds_size as usize));
+
+    // clone a esds cursor for parsing.
+    let esds = &mut Cursor::new(esds_array.clone());
+    let esds_tag = try!(esds.read_u8());
+
+    if esds_tag != ESDESCR_TAG {
+        return Err(Error::Unsupported("fail to parse ES descriptor"));
+    }
+
+    let esds_extend = try!(esds.read_u8());
+    // extension tag start from 0x80.
+    if esds_extend >= 0x80 {
+        // skip remains extension and length.
+        try!(skip(esds, 5));
+    } else {
+        try!(skip(esds, 2));
+    }
+
+    let esds_flags = try!(esds.read_u8());
+
+    // Stream dependnecy flag, first bit from left most.
+    if esds_flags & 0x80 > 0 {
+        // Skip uninteresting fields.
+        try!(skip(esds, 2));
+    }
 
-    Ok(AudioCodecSpecific::ES_Descriptor(esds))
+    // Url flag, second bit from left most.
+    if esds_flags & 0x40 > 0 {
+        // Skip uninteresting fields.
+        let skip_es_len = try!(esds.read_u8()) + 2;
```

Comment:
It'd be safer to make skip_es_len a usize and cast the result of the read before adding to avoid potential overflows.